nodebuilder/blobstream/service_test.go

```diff
@@ -0,0 +1,1087 @@
+}
+
+// createTestBlobTransaction creates a test blob transaction using a specific signer and a specific PFB size.
+// The size is in bytes.
+func createTestBlobTransaction(t *testing.T, signer *types.KeyringSigner, size int) (namespace.Namespace, *types.MsgPayForBlobs, *types.Blob, coretypes.Tx) {
+	addr, err := signer.GetSignerInfo().GetAddress()
+	require.NoError(t, err)
+
+	ns := namespace.RandomBlobNamespace()
+	msg, blob := blobfactory.RandMsgPayForBlobsWithNamespaceAndSigner(addr.String(), ns, size)
+	require.NoError(t, err)
+
+	builder := signer.NewTxBuilder()
+	stx, err := signer.BuildSignedTx(builder, msg)
+	require.NoError(t, err)
+	rawTx, err := encoding.MakeConfig(app.ModuleEncodingRegisters...).TxConfig.TxEncoder()(stx)
+	require.NoError(t, err)
+	cTx, err := coretypes.MarshalBlobTx(rawTx, blob)
+	require.NoError(t, err)
+	return ns, msg, blob, cTx
+}
+
+func TestShareToSubtreeRootProof(t *testing.T) {
+	var shares [][]byte
+	// generate some shares
+	for i := 0; i < 10; i++ {
+		shares = append(shares, bytes.Repeat([]byte{0x1}, appconsts.ShareSize))
+	}
+	// calculate the expected proof
+	subtreeRoot, expectedProofs := merkle.ProofsFromByteSlices(shares)
+
+	// calculate the actual proofs
+	var actualProofs []*ResultShareToSubtreeRootProof
+	for i := range shares {
+		proof, err := ProveShareToSubtreeRoot(shares, uint64(i))
+		require.NoError(t, err)
+		actualProofs = append(actualProofs, proof)
+	}
+
+	// compare the proofs and validate
+	for shareIndex, actualProof := range actualProofs {
+		t.Run(fmt.Sprintf("shareIndex=%d", shareIndex), func(t *testing.T) {
+			valid, err := actualProof.ShareToSubtreeRootProof.Verify(subtreeRoot, shares[shareIndex])
+			assert.NoError(t, err)
+			assert.True(t, valid)
+			assert.Equal(t, *expectedProofs[shareIndex], actualProof.ShareToSubtreeRootProof.Proof)
+		})
+	}
+}
+
+func TestSubtreeRootsToCommitmentProof(t *testing.T) {
+	rowRootSize := sha256.Size + 2*appconsts.NamespaceSize
+	var subtreeRoots [][]byte
+	// generate some subtreeRoots
+	for i := 0; i < 10; i++ {
+		subtreeRoots = append(subtreeRoots, bytes.Repeat([]byte{0x1}, rowRootSize))
+	}
+	// calculate the expected proof
+	shareCommitment, expectedProofs := merkle.ProofsFromByteSlices(subtreeRoots)
+
+	// calculate the actual proofs
+	var actualProofs []*ResultSubtreeRootToCommitmentProof
+	for i := range subtreeRoots {
+		proof, err := ProveSubtreeRootToCommitment(subtreeRoots, uint64(i))
+		require.NoError(t, err)
+		actualProofs = append(actualProofs, proof)
+	}
+
+	// compare the proofs and validate
+	for subtreeRootIndex, actualProof := range actualProofs {
+		t.Run(fmt.Sprintf("subtreeRootIndex=%d", subtreeRootIndex), func(t *testing.T) {
+			valid, err := actualProof.SubtreeRootToCommitmentProof.Verify(shareCommitment, subtreeRoots[subtreeRootIndex])
+			assert.NoError(t, err)
+			assert.True(t, valid)
+			assert.Equal(t, *expectedProofs[subtreeRootIndex], actualProof.SubtreeRootToCommitmentProof.Proof)
+		})
+	}
+}
+
```

Comment:
instead of using the generated mock. I decided to implement a custom mock here so that I can test a lot of combinations without having to worry about the mocked p2p network etc.

If you check the `TestProveCommitmentAllCombinations` test, you will see that it's kind of fuzzing tons of combinations.